check.go

```diff
@@ -83,3 +91,102 @@ func Check(root string, dh *DirectoryHierarchy, keywords []string) (*Result, err
 	}
 	return &result, nil
 }
+
+// TarCheck is the tar equivalent of checking a file hierarchy spec against a tar stream to
+// determine if files have been changed.
+func TarCheck(tarDH, dh *DirectoryHierarchy, keywords []string) (*Result, error) {
+	var result Result
+	var err error
+	var tarRoot *Entry
+
+	for _, e := range tarDH.Entries {
+		if e.Name == "." {
+			tarRoot = &e
+			break
+		}
+	}
+	tarRoot.Next = &Entry{
+		Name: "seen",
+		Type: CommentType,
+	}
+	curDir := tarRoot
+	creator := dhCreator{DH: dh}
+	sort.Sort(byPos(creator.DH.Entries))
+
+	var outOfTree bool
+	for i, e := range creator.DH.Entries {
+		switch e.Type {
+		case SpecialType:
+			if e.Name == "/set" {
+				creator.curSet = &creator.DH.Entries[i]
+			} else if e.Name == "/unset" {
+				creator.curSet = nil
+			}
+		case RelativeType, FullType:
+			if outOfTree {
+				return &result, fmt.Errorf("No parent node from %s", e.Path())
+			}
+			// TODO: handle the case where "." is not the first Entry to be found
+			tarEntry := curDir.Descend(e.Name)
+			if tarEntry == nil {
+				// if directory is missing, we are still in parent directory,
+				// so we just move on to the next file or directory to traverse
+				// through. This should be ok.
+				result.Missing = append(result.Missing, e)
+				continue
+			}
+
+			tarEntry.Next = &Entry{
+				Type: CommentType,
+				Name: "seen",
+			}
+
+			// expected values from file hierarchy spec
+			var kvs KeyVals
+			if creator.curSet != nil {
+				kvs = MergeSet(creator.curSet.Keywords, e.Keywords)
+			} else {
+				kvs = NewKeyVals(e.Keywords)
+			}
+
+			// actual
+			var tarkvs KeyVals
+			if tarEntry.Set != nil {
+				tarkvs = MergeSet(tarEntry.Set.Keywords, tarEntry.Keywords)
+			} else {
+				tarkvs = NewKeyVals(tarEntry.Keywords)
+			}
+
+			for _, kv := range kvs {
+				if _, ok := KeywordFuncs[kv.Keyword()]; !ok {
+					return nil, fmt.Errorf("Unknown keyword %q for file %q", kv.Keyword(), e.Path())
+				}
+				if keywords != nil && !inSlice(kv.Keyword(), keywords) {
+					continue
+				}
+				if tarkv := tarkvs.Has(kv.Keyword()); tarkv != emptyKV {
+					if string(tarkv) != string(kv) {
+						failure := Failure{Path: tarEntry.Path(), Keyword: kv.Keyword(), Expected: kv.Value(), Got: tarkv.Value()}
+						result.Failures = append(result.Failures, failure)
+					}
+				}
+			}
+			// Step into a directory
+			if tarEntry.Prev != nil {
+				curDir = tarEntry
+			}
+		case DotDotType:
+			if outOfTree {
+				return &result, fmt.Errorf("No parent node.")
+			}
+			curDir = curDir.Ascend()
+			if curDir == nil {
+				outOfTree = true
+			}
+		}
+	}
+	result.Extra = filter(tarRoot, func(e *Entry) bool {
```

Comment:
Yes. The ability to do checking of *new* objects is incredibly important for the usecase I'm thinking of right now (an OCI-compliant image builder that just uses runC and OCI spec files).